File: leetcode/maximumSubarraySum.py
class Solution:
    def maximumSubarraySum(self, nums: List[int], k: int) -> int:
        numPositions = {}
        n = len(nums)
        prefix = [0] * (n+1)
        res = -inf+1

        for i, v in enumerate(nums):
            print(i, v-k)
            prefix[i+1] = prefix[i] + nums[i]
            if (v-k) in numPositions:
                pi = numPositions[v-k]
                res = max(res, prefix[i+1] - prefix[pi])
            if (v+k) in numPositions:
                pi = numPositions[v+k]
                res = max(res, prefix[i+1] - prefix[pi])
            if v in numPositions and prefix[i+1] - prefix[numPositions[v]] < nums[i]:
                numPositions[v] = i
            if v not in numPositions:
                numPositions[v] = i
        return res if res != -inf+1 else 0

File: leetcode/maximumSubarraySum.go
package main

func max(i, j int) int {
	if i > j {
		return i
	} else {
		return j
	}
}

func maximumSubarraySum(nums []int, k int) int64 {
	numPos := make(map[int][]int)
	var prefix []int
	prefix = append(prefix, 0)
	res := -1 << 52
	for i, v := range nums {
		prefix = append(prefix, prefix[len(prefix)-1]+v)
		numPos[v] = append(numPos[v], i)

		for _, pi := range numPos[v-k] {
			res = max(res, prefix[i+1]-prefix[pi])
		}
		for _, pi := range numPos[v+k] {
			res = max(res, prefix[i+1]-prefix[pi])
		}
	}
	if res != (-1 << 52) {
		return int64(res)
	} else {
		return 0
	}
}
